libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Extensions/RestClient.swift

```diff
@@ -0,0 +1,432 @@
+import Foundation
+
+/*
+ RestClient.swift
+ SalesforceSDKCore
+ 
+ Created by Raj Rao on 9/24/19.
+ 
+ Copyright (c) 2019-present, salesforce.com, inc. All rights reserved.
+ 
+ Redistribution and use of this software in source and binary forms, with or without modification,
+ are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this list of conditions
+ and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright notice, this list of
+ conditions and the following disclaimer in the documentation and/or other materials provided
+ with the distribution.
+ * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
+ endorse or promote products derived from this software without specific prior written
+ permission of salesforce.com, inc.
+ 
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import Combine
+/// Errors that can be thrown while using RestClient
+public enum RestClientError: Error {
+    case ApiResponseIsEmpty
+    case ApiInvocationFailed(underlyingError: Error, urlResponse: URLResponse?)
+    case DecodingFailed(underlyingError: Error)
+    case JsonSerialization(underlyingError: Error)
+}
+
+public struct RestResponse {
+    
+    private static let emptyJsonDictionaryArrayResponse = [String:Any]()
+    private static let emptyJsonArrayResponse = [[String:Any]]()
+    private static let emptyStringResponse = ""
+    
+    private (set) var data: Data
+    private (set) var urlResponse: URLResponse
+    
+    /// Initializes the RestResponse with a Data object and URLResponse
+    /// - Parameter data: Response as raw Data
+    /// - Parameter urlResponse: URlResponse from endpoint
+    public init(data: Data, urlResponse: URLResponse) {
+        self.data = data
+        self.urlResponse = urlResponse
+    }
+    
+    /// Parse the response as a Json Dictionary.
+    public func asJson() throws -> Any {
+        do {
+            let jsonData = try JSONSerialization.jsonObject(with: data, options: [])
+            return jsonData
+        } catch let error {
```

Comment:
Just like `Java try-catch-throw` (reminds me of how this was a no-no in `Obj C` land except in special cases ;-)).

Reply:
Throw in swift is more performant than in objective-c.

> Error handling in Swift resembles exception handling in other languages, with the use of the try, catch and throw keywords. Unlike exception handling in many languages—including Objective-C—error handling in Swift does not involve unwinding the call stack, a process that can be computationally expensive. As such, the performance characteristics of a throw statement are comparable to those of a return statement.

Source: https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html